src/aiida/orm/fields.py

```diff
@@ -0,0 +1,297 @@
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Module which provides decorators for AiiDA ORM entity -> DB field mappings."""
+from abc import ABCMeta
+from copy import deepcopy
+from pprint import pformat
+from typing import Any, Dict, Iterable, Optional, Sequence, Tuple, Type, Union
+
+__all__ = ('QbAttrField', 'QbField', 'QbFields', 'QbFieldFilters')
+
+
+class QbField:
+    """A field of an ORM entity, accessible via the ``QueryBuilder``"""
+
+    __slots__ = ('_key', '_qb_field', '_doc', '_dtype', '_subscriptable')
+
+    def __init__(
+        self,
+        key: str,
+        qb_field: Optional[str] = None,
+        *,
+        dtype: Optional[Type[Any]] = None,
+        doc: str = '',
+        subscriptable: bool = False,
+    ) -> None:
+        """Initialise a ORM entity field, accessible via the ``QueryBuilder``
+
+        :param key: The key of the field on the ORM entity
+        :param qb_field: The name of the field in the QueryBuilder, if not equal to ``key``
+        :param dtype: The data type of the field. If None, the field is of variable type.
+        :param doc: A docstring for the field
+        :param subscriptable: If True, a new field can be created by ``field["subkey"]``
+        """
+        self._key = key
+        self._qb_field = qb_field if qb_field is not None else key
+        self._doc = doc
+        self._dtype = dtype
+        self._subscriptable = subscriptable
+
+    @property
+    def key(self) -> str:
+        return self._key
+
+    @property
+    def qb_field(self) -> str:
+        return self._qb_field
+
+    @property
+    def doc(self) -> str:
+        return self._doc
+
+    @property
+    def dtype(self) -> Optional[Type[Any]]:
+        return self._dtype
+
+    @property
+    def subscriptable(self) -> bool:
+        return self._subscriptable
+
+    def _repr_type(self, value: Any) -> str:
+        """Return a string representation of the type of the value"""
+        if value == type(None):
+            return 'None'
+        if isinstance(value, type):
+            # basic types like int, str, etc.
+            return value.__qualname__
+        if hasattr(value, '__origin__') and value.__origin__ == Union:
+            return 'typing.Union[' + ','.join(self._repr_type(t) for t in value.__args__) + ']'
+        return str(value)
+
+    def __repr__(self) -> str:
+        dtype = self._repr_type(self.dtype) if self.dtype else ''
+        return (
+            f'{self.__class__.__name__}({self.key!r}'
+            + (f', {self._qb_field!r}' if self._qb_field != self.key else '')
+            + (f', dtype={dtype}' if self.dtype else '')
+            + (f', subscriptable={self.subscriptable!r}' if self.subscriptable else '')
+            + ')'
+        )
+
+    def __str__(self) -> str:
+        type_str = (
```

Comment:
This if statement is upsetting `mypy`, specifically raising a `redundant-expr` error, as `if isinstance(<whatever>, type)` is always `True`. Or is it not? Can't quite tell what @chrisjsewell was aiming for with this.